poc/idpf_poplar.sage

```diff
@@ -127,18 +131,23 @@ class IdpfPoplar(Idpf):
     @classmethod
     def eval_next(IdpfPoplar, prev_seed, prev_ctrl,
                   correction_word, level, bit):
+        Field = IdpfPoplar.current_field(level)
         (seed_cw, ctrl_cw, w_cw) = correction_word
         (s, t) = IdpfPoplar.extend(prev_seed)
-        if prev_ctrl == Field2(1):
-            s[0] = xor(s[0], seed_cw)
-            s[1] = xor(s[1], seed_cw)
-            t[0] = t[0] + ctrl_cw[0]
-            t[1] = t[1] + ctrl_cw[1]
+        s[0] = xor(s[0], prev_ctrl.mask(seed_cw))
+        s[1] = xor(s[1], prev_ctrl.mask(seed_cw))
+        t[0] += ctrl_cw[0] * prev_ctrl
+        t[1] += ctrl_cw[1] * prev_ctrl
 
         next_ctrl = t[bit]
         (next_seed, y) = IdpfPoplar.convert(level, s[bit])
-        if next_ctrl == Field2(1):
-            y = vec_add(y, w_cw)
+        # Implementation note: Here we add the correction word to the
+        # output if `next_ctrl` is set. We avoid branching on the value of
+        # the control bit in order to reduce side channel leakage.
+        mask = Field(next_ctrl.as_unsigned())
+        for i in range(len(y)):
+            y[i] += w_cw[i] * mask
```

Comment:
In other languages, there might be a concern that the compiler "optimizes" this to `y[i] = 0 if mask == 0 else w_cw[i]` or something similar, in particular if checking for zero is cheaper than multiplication. Not sure if that's the case with Sage.

Reply:
Hmm yeah. Though I think this falls into the category of "finite field arithmetic needs to be constant time", i.e., https://github.com/divviup/libprio-rs/blob/main/src/fp.rs#L81.